### PR TITLE
Wait for main process in _save_checkpoint to ensure best checkpoint exists

### LINKED ISSUE #38008: Trainer Stuck at 0% Progress during Training on Multi-GPU Setup
### System Info

- transformers version: 4.37.2
- Platform: Linux version 5.15.0-105-generic
- Python version: 3.10.17
- Huggingface_hub version: 0.31.1
- Safetensors version: 0.5.3
- Accelerate version: 0.26.0
- PyTorch version (GPU?): 2.1.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Machine: 4 x A800 GPUs

### Who can help?

@ArthurZucker 
@zach-huggingface
@SunMarc

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I am running the following script:

> CUDA_VISIBLE_DEVICES=0,1,2,3` python svd_trainer.py --model_id meta-llama/Llama-2-7b-hf --target_ratio 0.4 --remapping

The core parts of the code that may be related to the issue are:

>  
   # training
    data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)
    from transformers import TrainingArguments
    training_args = TrainingArguments(
        output_dir= TA_tarined_model_output_dir,
        num_train_epochs = TA_num_train_epochs,
        evaluation_strategy = "epoch",
        per_device_train_batch_size = 1,
        per_device_eval_batch_size = 1,
        warmup_steps = TA_warmup_steps,
        lr_scheduler_type = "cosine",
        seed = args.seed,
        gradient_accumulation_steps = TA_gradient_accumulation_steps,
        # load_best_model_at_end = True,
        save_strategy = "no",
        save_steps = 1000,
        save_total_limit = 2,
        remove_unused_columns=False,
        # deepspeed =deepspeed_config
    )
   
    trainer = SVDTrainer(
        model=model,
        args=training_args,
        train_dataset=tokenized_traindata,
        eval_dataset=tokenized_valdata,
        data_collator=data_collator,
    )

    
    model,  train_dataloader, eval_dataloader = accelerator.prepare(
        model, trainer.get_train_dataloader(), trainer.get_eval_dataloader()
    )
    trainer.train = accelerator.prepare(trainer.train)
    
    
    # save setting
    TA_tarined_model_output_dir.mkdir(parents=True, exist_ok=True)
    output_json_path = TA_tarined_model_output_dir / "para_config.json"
    
    data = {
        "model_id": model_id,
        "DATASET_NAME": DATASET_NAME,
        "target_compression_ratio": target_compression_ratio,
        "BETA": BETA,
        "dataset_processing": {
            "SEED": args.seed,
            "SEQ_LEN": SEQ_LEN,
            "NSAMPLES_train": NSAMPLES_train,
            "NSAMPLES_val": NSAMPLES_val,
        },
        "training_settings": {
            "TA_num_train_epochs": TA_num_train_epochs,
            "TA_warmup_steps": TA_warmup_steps,
            "TA_gradient_accumulation_steps": TA_gradient_accumulation_steps
        },
        "loss_function_settings": {
            "lambda_reg": lambda_reg
        },
        "optimizer_settings": {
            "scheduler_lr": scheduler_lr,
            "scheduler_step_size": scheduler_step_size,
            "scheduler_gamma": scheduler_gamma
            # "scheduler_half_step_size": scheduler_half_step_size,
            # "scheduler_divide": scheduler_divide
        },
        "dtype_settings": {
            "model_load_dtype": "f16",
            "computeSVD_dtype": "f32"
        },
        "no_svd_layer": model_no_svd_layer_dic[lower_id],
    }
    with open(output_json_path, "w") as json_file:
        json.dump(data, json_file, indent=4)

    
    print("training start, target_compression_ratio is ",target_compression_ratio, " , NGPUS is ", NGPUS)
    trainer.train()
    
    accelerator.wait_for_everyone()

It stuck at 0% more than 12 hours.

Image

I wrote a test code to determine that Trainer can run on a single GPU but not on multiple GPUs. However, the program will report an error after disabling P2P by referring to the solution in https://github.com/huggingface/transformers/issues/26724.

Image

Here's how I checked the communication between the GPUs after disabling P2P

Image

### Expected behavior

If you have any insight or suggestions on how to resolve this issue, we would appreciate it.Thanks!

### PR DESCRIPTION
**Pull request by ssharpe42**
# What does this PR do?

In 4.50.0 a bug was introduced with a refactor of the best checkpoint process when we try to load the best model at end and  a non-main process has a null `self.state.best_model_checkpoint`. 

Before we run `_load_best_model()` there is a barrier to make sure the main process saves the model, however inside `_save_checkpoint`, this is missing. The model is saved [here](https://github.com/huggingface/transformers/blob/0b057e66b52556da3a1cbc29e2a98c0784ea9c33/src/transformers/trainer.py#L3197) and then a couple lines later the `self.state.best_model_checkpoint` is set only if the `best_checkpoint_dir` exists. When another process gets to this line first, `self.state.best_model_checkpoint` will be null. 

Finally, when training reaches the end and it is time to load the best model [here](https://github.com/huggingface/transformers/blob/0b057e66b52556da3a1cbc29e2a98c0784ea9c33/src/transformers/trainer.py#L2666) only the main process enters the if statement and the dist.barrier causes it to wait forever for process 1 to arrive which doesn't have a best_model_checkpoint.

In 4.48.3 it didn't check for existence before assigning the best model checkpoint. They refactored how this works, I can't get it to fail on a non-cube example yet. https://github.com/huggingface/transformers/blob/298b3f19303294293f7af075609481d64cb13de3/src/transformers/trainer.py#L3190 (edited)

Note: I didn't use `self.accelerator.wait_for_everyone()` since it doesn't seem to cover the sagemaker case, but I could make a wait function in the trainer since it is used twice now. 

I believe this fixes #38008

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@zach-huggingface, @SunMarc and @qgallouedec

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3346,6 +3346,15 @@ def _save_checkpoint(self, model, trial):
...
 
+        # Wait for everyone to get here so we are sure the model has been saved by process 0
+        # before we check if the best_checkpoint_dir exists
+        if is_torch_xla_available():
+            xm.rendezvous("load_best_model_at_end")
+        elif self.args.parallel_mode == ParallelMode.DISTRIBUTED:
+            dist.barrier()
+        elif is_sagemaker_mp_enabled():
+            smp.barrier()
+
```
Comment (SunMarc) [inline src/transformers/trainer.py]: let's put that in the condition below. Also maybe we can remove the following code as we are making sure here that the model was saved ? 
```python 
if args.load_best_model_at_end and self.state.best_model_checkpoint is not None:
            # Wait for everyone to get here so we are sure the model has been saved by process 0.
            if is_torch_xla_available():
                xm.rendezvous("load_best_model_at_end")
            elif args.parallel_mode == ParallelMode.DISTRIBUTED:
                dist.barrier()
            elif is_sagemaker_mp_enabled():
                smp.barrier()
```

**Review by SunMarc — COMMENTED** [review]
Thanks a lot, makes sense. Just a minor nit

**Review by SunMarc — APPROVED** [review]
Nice ! LGTM

**Comment by ssharpe42** [comment]
Should we rerun checks again? Trying to merge so we can get this in the next release.
